umi_tools/whitelist.py

```diff
@@ -0,0 +1,456 @@
+
+    # check whether the pattern string contains umi/cell bases
+    elif options.extract_method == "string":
+        if options.pattern:
+            if "C" in options.pattern:
+                extract_cell = True
+            if "N" in options.pattern:
+                extract_umi = True
+        if options.pattern2:
+            if "C" in options.pattern2:
+                extract_cell = True
+            if "N" in options.pattern2:
+                extract_umi = True
+
+    if not extract_umi:
+        if options.extract_method == "string":
+            U.error("barcode pattern(s) do not include any umi bases "
+                    "(marked with 'Ns') %s, %s" % (
+                        options.pattern, options.pattern2))
+        elif options.extract_method == "regex":
+            U.error("barcode regex(es) do not include any umi groups "
+                    "(starting with 'umi_') %s, %s" (
+                        options.pattern, options.pattern2))
+    if not extract_cell:
+        if options.extract_method == "string":
+            U.error("barcode pattern(s) do not include any cell bases "
+                    "(marked with 'Cs') %s, %s" % (
+                        options.pattern, options.pattern2))
+        elif options.extract_method == "regex":
+            U.error("barcode regex(es) do not include any cell groups "
+                    "(starting with 'cell_') %s, %s" (
+                        options.pattern, options.pattern2))
+
+    if options.stdin == sys.stdin:
+        read1s = umi_methods.fastqIterate(U.openFile(options.stdin))
+    else:
+        read1s = umi_methods.fastqIterate(U.openFile(options.stdin.name))
+
+    # set up read extractor
+    ReadExtractor = umi_methods.ExtractFilterAndUpdate(
+        method=options.extract_method,
+        pattern=options.pattern,
+        pattern2=options.pattern2,
+        prime3=options.prime3,
+        extract_cell=extract_cell)
+
+    cell_barcode_counts = collections.Counter()
+
+    n_reads = 0
+    n_cell_barcodes = 0
+
+    # if using the umis method, need to keep a set of umis observed
+    if options.method == "umis":
+        cell_barcode_umis = collections.defaultdict(set)
+
+    if not options.read2_in:
+        for read1 in read1s:
+            n_reads += 1
+            barcode_values = ReadExtractor.getBarcodes(read1)
+            if barcode_values is None:
+                continue
+            else:
+                cell, umi, _, _, _, _, _ = barcode_values
+                if options.method == "umis":
+                    cell_barcode_umis[cell].add(umi)
+                else:
+                    cell_barcode_counts[cell] += 1
+                n_cell_barcodes += 1
+
+            if options.subset_reads:
+                if n_cell_barcodes > options.subset_reads:
+                    break
+    else:
+        read2s = umi_methods.fastqIterate(U.openFile(options.read2_in))
+        for read1, read2 in izip(read1s, read2s):
```

Comment:
I was being silly. Just realised `joinedFastqIterate` is only used in `extract` and required to handle cases where the user has already filtered the reads for some reason and the reads are not strictly paired, as per the output from `alevin`. For `whitelist` we can reasonably demand that the user has not performed any pre-filtering as this would invalidate the assumptions used to define the whitelist so we can stick with the above `izip(read1s, read2s)`